README.md

## About the code
This nodejs code scrapes the HTTP Response Status Codes from the Mozilla Developer website. The final result is the creation of a module which enumerates the HTTP Status Codes. This module may be used in other JavaScript code and assists in complying with TypeScript rules regarding "Magic Numbers" (i.e. "HttpStatus.OK" versus 200).

### Reference Material
The Ultimate Guide to Web Scraping with Node.js</br>
*by Daniel Ni*</br>
https://www.freecodecamp.org/news/the-ultimate-guide-to-web-scraping-with-node-js-daa2027dcd3/

HTTP Response Status Codes</br>
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

Alternative nodejs code...</br>
*by prettymuchbryce*</br>
http-status-codes</br>
https://www.npmjs.com/package/http-status-codes

Alternative nodejs repo...</br>
*by prettymuchbryce*</br>
http-status-codes</br>
https://github.com/prettymuchbryce/http-status-codes#readme

### Nodejs modules
* request
* request-promise</br>
https://github.com/request/request-promise

* puppeteer</br>
https://github.com/GoogleChrome/puppeteer

* cheerio</br>
https://github.com/cheeriojs/cheerio

### Usage
In a terminal window, type:</br>
<strong>node generateHttpStatusCodes.js</strong>

### Version
#### v1.0 - HTTP Status Codes Scraper.</br>
The results after running the command above are as follows:
<div style="color:limegreen";>100 Continue</div>

#### v2.0 - Create aray of status codes.</br>
The results after running the command above are as follows:</br>
<div style="color:limegreen";>
[</br>
  { code: '100', response: 'Continue' },</br>
  { code: '101', response: 'Switching Protocol' },</br>
  { code: '102', response: 'Processing' },</br>
  { code: '103', response: 'Early Hints' },</br>
  { code: '200', response: 'OK' }, ...</br>
]
</div>

#### v3.0 - Create enumerated list for class module.
The results after running the command above are as follows:</br>
<div style="color:limegreen";>
this.CONTINUE = 100;</br>
this.SWITCHING_PROTOCOL = 101;</br>
this.PROCESSING = 102;</br>
this.EARLY_HINTS = 103;</br>
this.OK = 200; ...</br>
</div>

#### v4.0 - Add class module file to current project.

Note:</br>
To reference the file '*scripts/httpStatusCodes.js*' in your project 
you must add the following lines to a javascript file within your project:

* import { httpStatusCodes } from "./httpStatusCodes.js";</br>
* const HttpStatus = new httpStatusCodes;</br>

To use the enumerated list of status codes, simply type the constant and select a code from the list. 

Examples...
* HttpStatus.OK
* HttpStatus.UNAUTHORIZED

#### v5.0 - Resolve dependency vulnerabilities.

As Administrator, run: 
<div style="color:aqua";>npm audit fix</div>


CHANGELOG.md
# Changelog
All  changes to this project will be documented in this file.
<br/><br/>

## [Unreleased]
<br/><br/>


## [v5.0] - 2020-04-10

=== npm audit security report ===                        
                                                                                
### Run  npm update extract-zip --depth 2  to resolve 1 vulnerability
                                 
  Low             Prototype Pollution                                                                     
  Package         minimist                                                                      
  Dependency of   puppeteer                                                                     
  Path            puppeteer > extract-zip > mkdirp > minimist                   
                                   
  More info       https://npmjs.com/advisories/1179                             

### Run  npm update https-proxy-agent --depth 2  to resolve 1 vulnerability
                                   
  High            Machine-In-The-Middle                                                         
  Package         https-proxy-agent                                                             
  Dependency of   puppeteer                                                                     
  Path            puppeteer > https-proxy-agent                                                             
  More info       https://npmjs.com/advisories/1184                             

found 2 vulnerabilities (1 low, 1 high) in 155 scanned packages
  run `npm audit fix` to fix 2 of them.
<br/><br/>


## [v4.0] - 2019-10-28
### Add class module file to current project.
- Add class module file 'scripts/httpStatusCodes.js' to current project.
- Add usage notes in README for other projects.
<br/><br/>


## [v3.0] - 2019-10-28
### Create enumerated list for class module.
- Create enumerated list for use as class module in other projects.
<br/><br/>


## [v2.0] - 2019-10-28
### Create array of status codes.
- Use element referencing to find desired values from webpage.
<br/><br/>


## [v1.0] - 2019-10-28
### HTTP Status Codes Scraper.
- Web Scrape HTTP Status Codes from the Mozilla reference.
<br/><br/>

generateHttpStatusCodes.js
// The Ultimate Guide to Web Scraping with Node.js
// by Daniel Ni
// https://www.freecodecamp.org/news/the-ultimate-guide-to-web-scraping-with-node-js-daa2027dcd3/

// HTTP response status codes
// https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

// Alternative nodejs code...
// by prettymuchbryce
// http-status-codes
// https://www.npmjs.com/package/http-status-codes

// Alternative repo...
// by prettymuchbryce
// http-status-codes
// https://github.com/prettymuchbryce/http-status-codes#readme

const rp = require('request-promise');
const $ = require('cheerio');
const url = 'https://developer.mozilla.org/en-US/docs/Web/HTTP/Status';

rp(url)
  .then(function(html){
    //success!
    let statusCodes = $('dt > a > code', html);

    const arrayStatusCodes = [];
    for (let i=0; i < statusCodes.length; i++) {
      let objData = statusCodes[i].firstChild.nodeValue;
      let statusCode = objData.split(" ")[0];
      let statusResponse = objData.slice(statusCode.length).trimLeft();
      let responseCodeData = {
        code: statusCode,
        response: statusResponse
      };
      arrayStatusCodes.push(responseCodeData);

      let txtResponse = statusResponse.toUpperCase();
      txtResponse = txtResponse.replace(/-/g, "_"); // global replacement
      txtResponse = txtResponse.replace(/ /g, "_"); // global replacement
      console.log("this." + txtResponse + " = " + statusCode + ";");
    }

    // console.log(arrayStatusCodes);
  })
  .catch(function(err){
    //handle error
    console.log(err.message);
  });

scripts/httpStatusCodes.js
// List acquired from the following code...
// https://github.com/git4m2/HTTP_Status_Codes_Scraper

// HTTP response status codes
// https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

class httpStatusCodes {
    constructor() {
        this.CONTINUE = 100;
        this.SWITCHING_PROTOCOL = 101;
        this.PROCESSING = 102;
        this.EARLY_HINTS = 103;
        this.OK = 200;
        this.CREATED = 201;
        this.ACCEPTED = 202;
        this.NON_AUTHORITATIVE_INFORMATION = 203;
        this.NO_CONTENT = 204;
        this.RESET_CONTENT = 205;
        this.PARTIAL_CONTENT = 206;
        this.MULTI_STATUS = 207;
        this.ALREADY_REPORTED = 208;
        this.IM_USED = 226;
        this.MULTIPLE_CHOICE = 300;
        this.MOVED_PERMANENTLY = 301;
        this.FOUND = 302;
        this.SEE_OTHER = 303;
        this.NOT_MODIFIED = 304;
        this.TEMPORARY_REDIRECT = 307;
        this.PERMANENT_REDIRECT = 308;
        this.BAD_REQUEST = 400;
        this.UNAUTHORIZED = 401;
        this.PAYMENT_REQUIRED = 402;
        this.FORBIDDEN = 403;
        this.NOT_FOUND = 404;
        this.METHOD_NOT_ALLOWED = 405;
        this.NOT_ACCEPTABLE = 406;
        this.PROXY_AUTHENTICATION_REQUIRED = 407;
        this.REQUEST_TIMEOUT = 408;
        this.CONFLICT = 409;
        this.GONE = 410;
        this.LENGTH_REQUIRED = 411;
        this.PRECONDITION_FAILED = 412;
        this.PAYLOAD_TOO_LARGE = 413;
        this.URI_TOO_LONG = 414;
        this.UNSUPPORTED_MEDIA_TYPE = 415;
        this.REQUESTED_RANGE_NOT_SATISFIABLE = 416;
        this.EXPECTATION_FAILED = 417;
        this.I_AM_A_TEAPOT = 418;
        this.MISDIRECTED_REQUEST = 421;
        this.UNPROCESSABLE_ENTITY = 422;
        this.LOCKED = 423;
        this.FAILED_DEPENDENCY = 424;
        this.TOO_EARLY = 425;
        this.UPGRADE_REQUIRED = 426;
        this.PRECONDITION_REQUIRED = 428;
        this.TOO_MANY_REQUESTS = 429;
        this.REQUEST_HEADER_FIELDS_TOO_LARGE = 431;
        this.UNAVAILABLE_FOR_LEGAL_REASONS = 451;
        this.INTERNAL_SERVER_ERROR = 500;
        this.NOT_IMPLEMENTED = 501;
        this.BAD_GATEWAY = 502;
        this.SERVICE_UNAVAILABLE = 503;
        this.GATEWAY_TIMEOUT = 504;
        this.HTTP_VERSION_NOT_SUPPORTED = 505;
        this.VARIANT_ALSO_NEGOTIATES = 506;
        this.INSUFFICIENT_STORAGE = 507;
        this.LOOP_DETECTED = 508;
        this.NOT_EXTENDED = 510;
        this.NETWORK_AUTHENTICATION_REQUIRED = 511;
    }
}

export { httpStatusCodes };